[webkitpy] SimulatedDeviceManager.initialize_devices should rebuild the dyld shared cache before booting simulators with none already booted.

bls1999 opened the pull request:
#### 50140920094471b706bb200855128b5d25b6976a
<pre>
[webkitpy] SimulatedDeviceManager.initialize_devices should rebuild the dyld shared cache before booting simulators with none already booted.
<a href="https://bugs.webkit.org/show_bug.cgi?id=284494">https://bugs.webkit.org/show_bug.cgi?id=284494</a>
<a href="https://rdar.apple.com/141316056">rdar://141316056</a>

Reviewed by Jonathan Bedard.

If booted before the dyld shared cache has a chance to populate, simulator
processes under SimulatorTrampoline appear to load all of the relevant
frameworks into memory for each process. This is causing severe simulator
performance regressions, which slows down testing and eats into host device
swap memory (especially when using multiple simulators in parallel).

Anecdotal API test stats (running 1 simulator):
 - No dyld shared cache: 3:00:51
 - With dyld shared cache: 1:44:57

For this reason, if no simulators are booted, we should rebuild the dyld shared
cache before booting new ones.

* Tools/Scripts/webkitpy/xcode/simulated_device.py:
    -&gt; (SimulatedDeviceManager.Runtime.__init__): Add new properties.
    -&gt; (SimulatedDeviceManager.Runtime.rebuild_dyld_shared_cache): Add new function that rebuilds the shared cache for the runtime.
    -&gt; (SimulatedDeviceManager.initialize_devices): Call rebuild_dyld_shared_cache if no simulators are booted.
* Tools/Scripts/webkitpy/xcode/simulated_device_unittest.py: Adjusting unit test mock input from simctl json.

Canonical link: <a href="https://commits.webkit.org/287716@main">https://commits.webkit.org/287716@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ab38c3a535583e943e0c80577b5dad8a73b1e0d1

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/80598 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/59605 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/34531 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/85122 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/31580 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/82709 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/68666 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/7911 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/62957 "Found 27 new test failures: imported/w3c/web-platform-tests/css/css-text-decor/text-emphasis-position-under-left-001.xht imported/w3c/web-platform-tests/css/css-text-decor/text-emphasis-position-under-right-001.xht imported/w3c/web-platform-tests/css/css-text/hyphens/hyphens-auto-last-word-001.html imported/w3c/web-platform-tests/css/css-text/hyphens/hyphens-character.html imported/w3c/web-platform-tests/css/css-text/hyphens/hyphens-vs-float-clearance-002.html imported/w3c/web-platform-tests/css/css-transforms/animation/transform-non-invertible-discrete-interpolation.html imported/w3c/web-platform-tests/css/css-transforms/transform-box/cssbox-content-box-002.html imported/w3c/web-platform-tests/css/css-view-transitions/pseudo-with-classes-exit.html imported/w3c/web-platform-tests/css/css-view-transitions/pseudo-with-classes-match-ident.html imported/w3c/web-platform-tests/css/css-view-transitions/pseudo-with-classes-match-multiple.html ... (failure)") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/20760 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/83667 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/53068 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/73373 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/43262 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/80056 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/50362 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/27533 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/30039 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/71508 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/28056 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/86555 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/7824 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/5519 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/71256 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/7999 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/69210 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/70496 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/14504 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/13450 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12483 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/7786 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/7625 "Built successfully") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/11144 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/9430 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

JonWBedard commented on Tools/Scripts/webkitpy/xcode/simulated_device.py:
```diff
@@ -57,15 +59,23 @@ def __init__(self, device_type, use_booted_simulator=True, use_existing_simulato
+            self.identifier: str = runtime_dict['identifier']
+            self.name: str = runtime_dict['name']
+
+        def rebuild_dyld_shared_cache(self, host: Union[SystemHost, None]):
```
I don't think we should include this type hint here. Strictly speaking, I don't think it's actually correct. This could be a mock host or really, any object which kind of "looks" like a host object.

bls1999 replied:
Will remove.

JonWBedard commented on Tools/Scripts/webkitpy/xcode/simulated_device.py:
```diff
@@ -473,6 +483,13 @@ def initialize_devices(cls, requests, host=None, name_base='Managed', simulator_
+            # New simulators will load all system frameworks into memory if the dyld shared cache doesn't exist.
+            # To avoid the immense strain this causes on the host's memory usage, we rebuild the cache here.
+            if len(SimulatedDeviceManager.INITIALIZED_DEVICES) == 0:
+                for runtime in SimulatedDeviceManager.AVAILABLE_RUNTIMES:
```
We need to keep track of whether we've run this command. We don't want to run it twice in the same test run.

Also, is this a one-time-per-Xcode install thing? I'm actually a little surprised we need to do this every single run.

bls1999 replied:
> We need to keep track of whether we've run this command. We don't want to run it twice in the same test run.

Good point, I will add that.

> Also, is this a one-time-per-Xcode install thing? I'm actually a little surprised we need to do this every single run.

It's supposed to be. Anecdotal evidence indicates the shared cache can get into a bad state, which causes it not to be able to be used. Better to run it each time to be as safe as possible.

JonWBedard commented on Tools/Scripts/webkitpy/xcode/simulated_device.py:
```diff
@@ -57,15 +59,26 @@ def __init__(self, device_type, use_booted_simulator=True, use_existing_simulato
+            self.version: str = Version.from_string(runtime_dict['version'])
+            self.identifier: str = runtime_dict['identifier']
+            self.name: str = runtime_dict['name']
+            self.rebuilt_cache: bool = False
```
Nit: Should probably be `has_rebuilt_cache`

bls1999 replied:
Done

bls1999 committed:
[webkitpy] SimulatedDeviceManager.initialize_devices should rebuild the dyld shared cache before booting simulators with none already booted.
https://bugs.webkit.org/show_bug.cgi?id=284494
rdar://141316056

Reviewed by Jonathan Bedard.

If booted before the dyld shared cache has a chance to populate, simulator
processes under SimulatorTrampoline appear to load all of the relevant
frameworks into memory for each process. This is causing severe simulator
performance regressions, which slows down testing and eats into host device
swap memory (especially when using multiple simulators in parallel).

Anecdotal API test stats (running 1 simulator):
 - No dyld shared cache: 3:00:51
 - With dyld shared cache: 1:44:57

For this reason, if no simulators are booted, we should rebuild the dyld shared
cache before booting new ones.

* Tools/Scripts/webkitpy/xcode/simulated_device.py:
    -> (SimulatedDeviceManager.Runtime.__init__): Add new properties.
    -> (SimulatedDeviceManager.Runtime.rebuild_dyld_shared_cache): Add new function that rebuilds the shared cache for the runtime.
    -> (SimulatedDeviceManager.initialize_devices): Call rebuild_dyld_shared_cache if no simulators are booted.
* Tools/Scripts/webkitpy/xcode/simulated_device_unittest.py: Adjusting unit test mock input from simctl json.

Canonical link: https://commits.webkit.org/287716@main
